src/graphql/data/TokenPrice.ts

```diff
@@ -1,84 +1,60 @@
-          timestamp
-          value
+        priceHistory(duration: $duration) {
+          timestamp @required(action: LOG)
+          value @required(action: LOG)
         }
       }
     }
   }
 `
-
-export type PricePoint = { timestamp: number; value: number }
+export type PricePoint = NonNullable<
+  NonNullable<NonNullable<NonNullable<TokenPriceQuery['response']['tokens']>[number]>['market']['priceHistory']>[number]
+>
 export type PriceDurations = Partial<Record<TimePeriod, PricePoint[]>>
 
-export function isPricePoint(p: { timestamp: number; value: number | null } | null): p is PricePoint {
-  return Boolean(p && p.value)
+export function isPricePoint(p: { timestamp: number; value: number } | null): p is PricePoint {
+  return p !== null
 }
 
-export function useTokenPriceQuery(address: string, chain: Chain): PriceDurations | undefined {
+export function useLoadTokenPriceQuery(address: string, chain: Chain, timePeriod: TimePeriod) {
+  const duration = toHistoryDuration(timePeriod)
   const contract = useMemo(() => ({ address: address.toLowerCase(), chain }), [address, chain])
-  const [prices, setPrices] = useState<PriceDurations>()
+  const queryReference = loadQuery<TokenPriceQuery>(useRelayEnvironment(), tokenPriceQuery, { contract, duration })
 
   useEffect(() => {
-    const subscription = fetchQuery<TokenPriceQuery>(environment, tokenPriceQuery, { contract }).subscribe({
-      next: (response: TokenPriceQuery['response']) => {
-        const priceData = response.tokens?.[0]?.market
-        const prices = {
-          [TimePeriod.HOUR]: priceData?.priceHistory1H?.filter(isPricePoint),
-          [TimePeriod.DAY]: priceData?.priceHistory1D?.filter(isPricePoint),
-          [TimePeriod.WEEK]: priceData?.priceHistory1W?.filter(isPricePoint),
-          [TimePeriod.MONTH]: priceData?.priceHistory1M?.filter(isPricePoint),
-          [TimePeriod.YEAR]: priceData?.priceHistory1Y?.filter(isPricePoint),
-        }
+    return () => queryReference.dispose()
+  })
 
-        // Ensure the latest price available is available for every TimePeriod.
-        const latests = Object.values(prices)
-          .map((prices) => prices?.slice(-1)?.[0] ?? null)
-          .filter(isPricePoint)
-        if (latests.length) {
-          const latest = latests.reduce((latest, pricePoint) =>
-            latest.timestamp > pricePoint.timestamp ? latest : pricePoint
-          )
-          Object.values(prices)
-            .filter((prices) => prices && prices.slice(-1)[0] !== latest)
-            .forEach((prices) => prices?.push(latest))
-        }
+  return queryReference
+}
+
+export function usePreloadedTokenPriceQuery(queryReference: PreloadedQuery<TokenPriceQuery>): PricePoint[] | undefined {
+  const queryData = usePreloadedQuery<TokenPriceQuery>(tokenPriceQuery, queryReference)
+
+  const priceHistory = useMemo(() => {
+    const priceHistory = queryData.tokens?.[0]?.market?.priceHistory?.filter(isPricePoint)
```

Comment:
To make this clearer, pull out the shared variable. Also, avoid mutating the query data by spreading the priceHistory.

    const market = queryData.tokens?.[0]?.market
    const priceHistory = market?.filter(isPricePoint)
    const currentPrice = market?.price?.value
    if (Array.isArray(priceHistory) && currentPrice !== undefined) {
      const timestamp = Date.now() / 1000
      return [...priceHistory, { timestamp, value: currentPrice })
    }
    return priceHistory